src/Strings.h

```diff
@@ -0,0 +1,388 @@
+
+    std::string trimedString(TrimAndReplace(input.begin(), input.end()));
+
+    if (ngramRangeMin < 1 || ngramRangeMin > ngramRangeMax || ngramRangeMax > trimedString.size())
+        throw std::invalid_argument("ngramRangeMin and ngramRangeMax not valid");
+
+    ParseNgramCharHelper<IteratorT>(trimedString.begin(), trimedString.end(), ngramRangeMin, ngramRangeMax, callback);
+}
+
+template <
+    typename IteratorT,
+    typename IsDelimiterT
+>
+inline void ParseNgramCharwb(std::string &input,
+                             IsDelimiterT const &isDelimiter,
+                             size_t const &ngramRangeMin,
+                             size_t const &ngramRangeMax,
+                             std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    std::string trimedString(TrimAndReplace(input.begin(), input.end()));
+
+    //wordIterPairVector is used to store the begin and end iterator of words in input
+    std::vector<std::pair<IteratorT, IteratorT>> wordIterPairVector;
+    Details::Parse<IteratorT, IsDelimiterT>(
+        trimedString.begin(),
+        trimedString.end(),
+        isDelimiter,
+        [&wordIterPairVector] (IteratorT &iterBegin, IteratorT &iterEnd) {
+            wordIterPairVector.emplace_back(std::make_pair(iterBegin, iterEnd));
+        }
+    );
+
+    if (wordIterPairVector.size() == 0)
+        throw std::invalid_argument("wordIterPairVector.size() == 0");
+
+    if (ngramRangeMin < 1 || ngramRangeMin > ngramRangeMax )
+        throw std::invalid_argument("ngramRangeMin and ngramRangeMax not valid");
+
+    for (size_t pairIdx = 0; pairIdx < wordIterPairVector.size(); ++pairIdx) {
+        //copy
+        std::string const paddedWord = " " + std::string(wordIterPairVector[pairIdx].first, wordIterPairVector[pairIdx].second) + " ";
+
+        ParseNgramCharHelper<IteratorT>(paddedWord.begin(), paddedWord.end(), ngramRangeMin, ngramRangeMax, callback);
+    }
+}
+
+}
+
+template <typename IteratorT>
+inline std::string ToLower(IteratorT begin,
+                           IteratorT end) {
+    std::string result(begin, end);
+
+#if (defined _MSC_VER)
+#   pragma warning(push)
+#   pragma warning(disable: 4244) // conversion from 'unsigned int' to 'unsigned char', possible loss of data
+#endif
+
+    std::transform(result.begin(), result.end(), result.begin(), [](char c) { return std::tolower(c); });
+
+#if (defined _MSC_VER)
+#   pragma warning(pop)
+#endif
+
+    return result;
+}
+
+template <typename IteratorT>
+inline std::string ToUpper(IteratorT begin,
+                           IteratorT end) {
+    std::string result(begin, end);
+
+#if (defined _MSC_VER)
+#   pragma warning(push)
+#   pragma warning(disable: 4244) // conversion from 'unsigned int' to 'unsigned char', possible loss of data
+#endif
+
+    std::transform(result.begin(), result.end(), result.begin(), [](char c) { return std::toupper(c); });
```

Comment:
>return std::toupper(c); [](start = 78, length = 23)

This handles only ASCII characters. You need to handle UTF-8.

Reply:
Yes this version only support ASCII characters, I'll add a todo here and will resolve in a future PR